From adc51752b776d655cf610950d7369718df324749 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Mon, 22 Oct 2018 19:13:15 +0800
Subject: [PATCH 4907/5242] MGS-4338-2 [#imx-1223] GPU memory optimization for
 page fault

commit  231cf7f48e099997e4ebb2072525d762c52e2045 from
https://source.codeaurora.org/external/imx/linux-imx.git

GPU memory initialization will cause large boot time with 40ms.

If GPU memory is allocated from CMA, memset in cma kernel driver,
It is not necessary to add duplicated memset in GPU kernel driver.

Only GFP allocator need trigger a page fault with a simple write.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../default/gc_hal_kernel_allocator_gfp.c          |    6 ++++++
 .../gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c |    3 ---
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_gfp.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_gfp.c
index 937c752..71a867c 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_gfp.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_gfp.c
@@ -926,6 +926,12 @@ static int gc_usage_show(struct seq_file* m, void* data)
 
     addr = vmap(pages, numPages, 0, pgprot_writecombine(PAGE_KERNEL));
 
+    /* Trigger a page fault. */
+    for (i = 0; i < numPages; i++)
+    {
+        *(gctINT *)(addr + PAGE_SIZE * i) = 0;
+    }
+
     if (free)
     {
         kfree(pages);
diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
index bc316b3..7be6e79 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_os.c
@@ -1345,9 +1345,6 @@
 
     gcmkONERROR(allocator->ops->MapKernel(allocator, mdl, &addr));
 
-    /* Trigger a page fault. */
-    memset(addr, 0, numPages * PAGE_SIZE);
-
     mdl->addr = addr;
 
     if (InUserSpace)
-- 
1.7.9.5

